File: src/core/actions.ts
/**
 * Acciones del jugador.
 *
 * Regla de diseño que corrige el problema central del juego anterior: **toda
 * acción da algo y cobra algo**. Antes, alimentar y jugar eran botones que
 * subían barras sin contrapartida, así que no había ninguna decisión que tomar
 * — solo clickear.
 *
 * Acá comer de más pasa factura, jugar gasta energía, y el vínculo tiene tope
 * diario para que la constancia valga más que el spam.
 */

import type { FoodKind } from "./evolution.ts";
import type { CreatureState } from "./simulation.ts";
import { localDayIndex } from "./simulation.ts";

export type { FoodKind };

export interface Food {
  id: string;
  name: string;
  /** Artículo, para poder armar frases sin que se rompa la concordancia. */
  articulo: "el" | "la";
  energia: number;
  animo: number;
  salud: number;
  /** Alimenta el vector de crianza que en la Fase 3 decide la rama evolutiva. */
  tipo: FoodKind;
}

export const FOODS: readonly Food[] = [
  { id: "baya", name: "Baya", articulo: "la", energia: 18, animo: 5, salud: 0, tipo: "dulce" },
  { id: "raiz", name: "Raíz", articulo: "la", energia: 26, animo: 0, salud: 1, tipo: "mineral" },
  {
    id: "larva",
    name: "Larva",
    articulo: "la",
    energia: 34,
    animo: -2,
    salud: 2,
    tipo: "proteina",
  },
  {
    id: "cristal",
    name: "Cristal",
    articulo: "el",
    energia: 12,
    animo: 12,
    salud: 3,
    tipo: "raro",
  },
];

/** A partir de acá, seguir comiendo rinde poco y empieza a hacer mal. */
const FULL_THRESHOLD = 85;
const OVERFEED_EFFICIENCY = 0.25;
const OVERFEED_HEALTH_COST = 2.5;

const PLAY_ENERGY_COST = 12;
const PLAY_MIN_ENERGY = 15;
const PLAY_MOOD_GAIN = 16;

const PET_MOOD_GAIN = 4;

const BOND_PER_ACTION = 2;
/**
 * Tope diario de vínculo.
 *
 * Es la regla que convierte "abrir la app" en un hábito en vez de en una sesión
 * de farmeo: alcanzado el tope, seguir interactuando no suma más. Premia volver
 * todos los días, no clickear cien veces hoy.
 */
export const BOND_DAILY_CAP = 12;

/** Salir del letargo cuesta vínculo. Se recupera la criatura, no el progreso. */
const LETHARGY_BOND_PENALTY = 0.25;

export type ActionResult =
  | { ok: true; state: CreatureState; message: string }
  | { ok: false; reason: string };

function clamp01to100(value: number): number {
  return value < 0 ? 0 : value > 100 ? 100 : value;
}

/** Copia el estado y aplica lo común a toda interacción. */
function touch(state: CreatureState, nowMs: number): { next: CreatureState; woke: boolean } {
  // Copia profunda: `crianza.dieta` es un objeto anidado y compartirlo por
  // referencia haría que la acción mutara el estado que recibió.
  const next: CreatureState = {
    ...state,
    stats: { ...state.stats },
    crianza: { ...state.crianza, dieta: { ...state.crianza.dieta } },
  };

  // El tope de vínculo se reinicia por día local, igual que en el tick.
  const day = localDayIndex(nowMs, next.tzOffsetMin);
  if (day !== next.diaIndice) {
    next.diaIndice = day;
    next.vinculoHoy = 0;
  }

  const woke = next.letargico;
  if (woke) {
    next.letargico = false;
    next.stats.vinculo = Math.max(0, next.stats.vinculo * (1 - LETHARGY_BOND_PENALTY));
  }
  next.ticksSinCuidado = 0;

  return { next, woke };
}

/** Suma vínculo respetando el tope diario. Devuelve cuánto entró de verdad. */
function grantBond(state: CreatureState, amount: number): number {
  const room = Math.max(0, BOND_DAILY_CAP - state.vinculoHoy);
  const granted = Math.min(amount, room);
  state.vinculoHoy += granted;
  state.stats.vinculo += granted;
  return granted;
}

function withWakeNote(message: string, woke: boolean): string {
  return woke ? `${message} Salió del letargo, pero el vínculo quedó golpeado.` : message;
}

export function alimentar(state: CreatureState, foodId: string, nowMs: number): ActionResult {
  const food = FOODS.find((item) => item.id === foodId);
  if (!food) return { ok: false, reason: `No existe el alimento "${foodId}"` };

  const { next, woke } = touch(state, nowMs);

  // Comer de más: rinde una cuarta parte y encima hace mal. Sin esto,
  // alimentar sería siempre la jugada correcta y no habría decisión.
  const full = next.stats.energia >= FULL_THRESHOLD;
  const gained = full ? food.energia * OVERFEED_EFFICIENCY : food.energia;

  next.stats.energia = clamp01to100(next.stats.energia + gained);
  next.stats.animo = clamp01to100(next.stats.animo + food.animo);
  next.stats.salud = clamp01to100(
    next.stats.salud + food.salud - (full ? OVERFEED_HEALTH_COST : 0),
  );
  grantBond(next, BOND_PER_ACTION);
  // La dieta se registra siempre, aun cuando comió sin ganas: lo que le diste
  // moldea en qué se convierte, más allá de cuánto le rindió esta vez.
  next.crianza.dieta[food.tipo]++;

  const comida = `${food.articulo} ${food.name.toLowerCase()}`;
  const message = full
    ? `Picoteó ${comida} sin ganas. Ya estaba llena.`
    : `Se morfó ${comida} sin respirar.`;

  return { ok: true, state: next, message: withWakeNote(message, woke) };
}

export function jugar(state: CreatureState, nowMs: number): ActionResult {
  if (state.stats.energia < PLAY_MIN_ENERGY) {
    return { ok: false, reason: "No le da la energía para jugar. Primero tiene que comer algo." };
  }

  const { next, woke } = touch(state, nowMs);

  // Jugar decaído rinde la mitad: la salud baja se nota en todo.
  const efficiency = next.stats.salud < 30 ? 0.5 : 1;
  next.stats.animo = clamp01to100(next.stats.animo + PLAY_MOOD_GAIN * efficiency);
  next.stats.energia = clamp01to100(next.stats.energia - PLAY_ENERGY_COST);
  grantBond(next, BOND_PER_ACTION);
  next.crianza.juego++;

  const message =
    efficiency < 1
      ? "Jugó un rato pero se cansó enseguida, pobre."
      : "Jugó hasta quedar rendida de contenta.";

  return { ok: true, state: next, message: withWakeNote(message, woke) };
}

export function acariciar(state: CreatureState, nowMs: number): ActionResult {
  const { next, woke } = touch(state, nowMs);

  next.stats.animo = clamp01to100(next.stats.animo + PET_MOOD_GAIN);
  const granted = grantBond(next, BOND_PER_ACTION);
  // Cuenta como crianza aunque el vínculo ya esté topeado: el tope limita el
  // vínculo, no el hecho de haber estado ahí.
  next.crianza.calma++;

  // Cuando el tope ya está alcanzado se dice, en vez de fingir que sumó.
  const message =
    granted > 0
      ? "Se dejó hacer mimos un buen rato."
      : "Está a gusto, pero por hoy ya tuvo lo suyo.";

  return { ok: true, state: next, message: withWakeNote(message, woke) };
}

File: gdext/src/actions.cpp
/**
 * actions.cpp — ver actions.h.
 */

#include "actions.h"

#include <algorithm>

namespace petbits {

const std::array<Food, 4> FOODS = {{
    {"baya", "Baya", "la", "baya", 18, 5, 0, FoodKind::Dulce},
    {"raiz", "Raíz", "la", "raíz", 26, 0, 1, FoodKind::Mineral},
    {"larva", "Larva", "la", "larva", 34, -2, 2, FoodKind::Proteina},
    {"cristal", "Cristal", "el", "cristal", 12, 12, 3, FoodKind::Raro},
}};

// El nombre en minúscula va precalculado en la tabla y no se baja de caso en
// tiempo de ejecución. Pasar "Raíz" a minúscula requiere saber que la Í con
// tilde ocupa dos bytes en UTF-8: std::tolower sobre bytes la rompe, y hacerlo
// bien pide una tabla Unicode para resolver un problema que no existe si el
// texto ya está escrito de las dos formas.

const Food* buscarAlimento(std::string_view id) {
    for (const Food& f : FOODS) {
        if (f.id == id) return &f;
    }
    return nullptr;
}

/** A partir de acá, seguir comiendo rinde poco y empieza a hacer mal. */
static constexpr double FULL_THRESHOLD = 85.0;
static constexpr double OVERFEED_EFFICIENCY = 0.25;
static constexpr double OVERFEED_HEALTH_COST = 2.5;

static constexpr double PLAY_ENERGY_COST = 12.0;
static constexpr double PLAY_MIN_ENERGY = 15.0;
static constexpr double PLAY_MOOD_GAIN = 16.0;

static constexpr double PET_MOOD_GAIN = 4.0;

static constexpr double BOND_PER_ACTION = 2.0;

/** Salir del letargo cuesta vínculo. Se recupera la criatura, no el progreso. */
static constexpr double LETHARGY_BOND_PENALTY = 0.25;

static double clamp01to100(double value) {
    return value < 0.0 ? 0.0 : (value > 100.0 ? 100.0 : value);
}

/**
 * ¿Está de expedición?
 *
 * De expedición no se la puede alimentar ni jugar con ella: no está. Se chequea
 * en cada acción y no dentro de `touch`, para poder devolver el motivo.
 */
static bool ausente(const CreatureState& state) {
    return state.expedicion.has_value();
}

static ActionResult rechazo(std::string motivo) {
    ActionResult r;
    r.ok = false;
    r.message = std::move(motivo);
    return r;
}

/** Copia el estado y aplica lo común a toda interacción. */
static void touch(CreatureState& next, int64_t nowMs, bool& woke) {
    // El tope de vínculo se reinicia por día local, igual que en el tick.
    const int64_t dia = localDayIndex(nowMs, next.tzOffsetMin);
    if (dia != next.diaIndice) {
        next.diaIndice = dia;
        next.vinculoHoy = 0.0;
    }

    woke = next.letargico;
    if (woke) {
        next.letargico = false;
        next.stats.vinculo = std::max(0.0, next.stats.vinculo * (1.0 - LETHARGY_BOND_PENALTY));
    }
    next.ticksSinCuidado = 0;
}

/** Suma vínculo respetando el tope diario. Devuelve cuánto entró de verdad. */
static double grantBond(CreatureState& state, double amount) {
    const double lugar = std::max(0.0, BOND_DAILY_CAP - state.vinculoHoy);
    const double dado = std::min(amount, lugar);
    state.vinculoHoy += dado;
    state.stats.vinculo += dado;
    return dado;
}

static std::string conNotaDeDespertar(std::string mensaje, bool woke) {
    if (!woke) return mensaje;
    return mensaje + " Salió del letargo, pero el vínculo quedó golpeado.";
}

/** Suma uno al contador de dieta que corresponda. */
static void anotarDieta(Crianza& c, FoodKind tipo) {
    switch (tipo) {
        case FoodKind::Proteina: c.dieta_proteina++; break;
        case FoodKind::Dulce:    c.dieta_dulce++;    break;
        case FoodKind::Mineral:  c.dieta_mineral++;  break;
        case FoodKind::Raro:     c.dieta_raro++;     break;
    }
}

// ---------------------------------------------------------------------------

ActionResult alimentar(const CreatureState& state, std::string_view foodId, int64_t nowMs) {
    if (ausente(state)) return rechazo("Está de expedición. Volvé cuando regrese.");

    const Food* food = buscarAlimento(foodId);
    if (food == nullptr) {
        return rechazo("No existe el alimento \"" + std::string(foodId) + "\"");
    }

    CreatureState next = state;
    bool woke = false;
    touch(next, nowMs, woke);

    // Comer de más: rinde una cuarta parte y encima hace mal.
    const bool llena = next.stats.energia >= FULL_THRESHOLD;
    const double ganado = llena ? food->energia * OVERFEED_EFFICIENCY : food->energia;

    next.stats.energia = clamp01to100(next.stats.energia + ganado);
    next.stats.animo = clamp01to100(next.stats.animo + food->animo);
    next.stats.salud =
        clamp01to100(next.stats.salud + food->salud - (llena ? OVERFEED_HEALTH_COST : 0.0));
    grantBond(next, BOND_PER_ACTION);
    // La dieta se registra siempre, aun cuando comió sin ganas: lo que le diste
    // moldea en qué se convierte, más allá de cuánto le rindió esta vez.
    anotarDieta(next.crianza, food->tipo);

    const std::string comida = std::string(food->articulo) + " " + std::string(food->nombreMinuscula);
    const std::string mensaje = llena ? "Picoteó " + comida + " sin ganas. Ya estaba llena."
                                      : "Se morfó " + comida + " sin respirar.";

    ActionResult r;
    r.ok = true;
    r.state = std::move(next);
    r.message = conNotaDeDespertar(mensaje, woke);
    return r;
}

ActionResult jugar(const CreatureState& state, int64_t nowMs) {
    if (ausente(state)) return rechazo("Está de expedición. Volvé cuando regrese.");

    if (state.stats.energia < PLAY_MIN_ENERGY) {
        return rechazo("No le da la energía para jugar. Primero tiene que comer algo.");
    }

    CreatureState next = state;
    bool woke = false;
    touch(next, nowMs, woke);

    // Jugar decaído rinde la mitad: la salud baja se nota en todo.
    const double eficiencia = next.stats.salud < 30.0 ? 0.5 : 1.0;
    next.stats.animo = clamp01to100(next.stats.animo + PLAY_MOOD_GAIN * eficiencia);
    next.stats.energia = clamp01to100(next.stats.energia - PLAY_ENERGY_COST);
    grantBond(next, BOND_PER_ACTION);
    next.crianza.juego++;

    const std::string mensaje = eficiencia < 1.0
                                    ? "Jugó un rato pero se cansó enseguida, pobre."
                                    : "Jugó hasta quedar rendida de contenta.";

    ActionResult r;
    r.ok = true;
    r.state = std::move(next);
    r.message = conNotaDeDespertar(mensaje, woke);
    return r;
}

ActionResult acariciar(const CreatureState& state, int64_t nowMs) {
    if (ausente(state)) return rechazo("Está de expedición. Volvé cuando regrese.");

    CreatureState next = state;
    bool woke = false;
    touch(next, nowMs, woke);

    next.stats.animo = clamp01to100(next.stats.animo + PET_MOOD_GAIN);
    const double dado = grantBond(next, BOND_PER_ACTION);
    // Cuenta como crianza aunque el vínculo ya esté topeado: el tope limita el
    // vínculo, no el hecho de haber estado ahí.
    next.crianza.calma++;

    // Cuando el tope ya está alcanzado se dice, en vez de fingir que sumó.
    const std::string mensaje = dado > 0.0 ? "Se dejó hacer mimos un buen rato."
                                           : "Está a gusto, pero por hoy ya tuvo lo suyo.";

    ActionResult r;
    r.ok = true;
    r.state = std::move(next);
    r.message = conNotaDeDespertar(mensaje, woke);
    return r;
}

} // namespace petbits
